x/protorev/keeper/rebalance.go

```diff
@@ -167,30 +171,174 @@ func (k Keeper) FindMaxProfitForRoute(ctx sdk.Context, route RouteMetaData, rema
+		if err != nil {
+			return sdk.ZeroInt(), err
+		}
+
+		tokenInDenom := inputDenom
+		if index > 0 {
+			tokenInDenom = route.Route[index-1].TokenOutDenom
+		}
+
+		switch {
+		case pool.GetType() == poolmanagertypes.Concentrated:
+			// If the pool is a concentrated liquidity pool, then check the maximum amount in that can be used
+			// and determine what this amount is as an input at the previous pool (working all the way up to the
+			// beginning of the route).
+			maxTokenIn, maxTokenOut, err := k.concentratedLiquidityKeeper.ComputeMaxInAmtGivenMaxTicksCrossed(
+				ctx,
+				pool.GetId(),
+				tokenInDenom,
+				types.MaxTicksCrossed,
+			)
+			if err != nil {
+				return sdk.ZeroInt(), err
+			}
+
+			// if there have been no other CL pools in the route, then we can set the intermediate coin to the max input amount.
+			// Additionally, if the amount of the previous token is greater than the possible amount from this pool, then we
+			// can set the intermediate coin to the max input amount (from the current pool). Otherwise we have to do a
+			// safe swap given the previous max amount.
+			if intermidiateCoin.IsNil() || maxTokenOut.Amount.LT(intermidiateCoin.Amount) {
+				intermidiateCoin = maxTokenIn
+				continue
+			}
+
+			// In the scenario where there are multiple CL pools in a route, we select the one that inputs
+			// the smaller amount to ensure we do not overstep the max ticks moved.
+			intermidiateCoin, err = k.executeSafeSwap(ctx, pool.GetId(), intermidiateCoin, tokenInDenom, route.StepSize)
+			if err != nil {
+				return sdk.ZeroInt(), err
+			}
+		case !intermidiateCoin.IsNil():
+			// If we have already seen a CL pool in the route, then simply propagate the intermediate coin up
+			// the route.
+			intermidiateCoin, err = k.executeSafeSwap(ctx, pool.GetId(), intermidiateCoin, tokenInDenom, route.StepSize)
+			if err != nil {
+				return sdk.ZeroInt(), err
+			}
+		}
+	}
+
+	// In the case where there are no CL pools, we want to return the extended max input amount
+	if intermidiateCoin.IsNil() {
+		return types.ExtendedMaxInputAmount, nil
+	}
+
+	// Ensure that the normalized upper bound is not greater than the extended max input amount
+	upperBound := intermidiateCoin.Amount.Quo(route.StepSize)
+	if upperBound.GT(types.ExtendedMaxInputAmount) {
+		return types.ExtendedMaxInputAmount, nil
+	}
+
+	return upperBound, nil
+}
+
+// executeSafeSwap executes a safe swap by first ensuring the swap amount is less than the
+// amount of total liquidity in the pool.
+func (k Keeper) executeSafeSwap(
+	ctx sdk.Context,
+	poolID uint64,
+	outputCoin sdk.Coin,
+	tokenInDenom string,
+	stepSize sdk.Int,
+) (sdk.Coin, error) {
+	liquidity, err := k.poolmanagerKeeper.GetTotalPoolLiquidity(ctx, poolID)
+	if err != nil {
+		return sdk.NewCoin(tokenInDenom, sdk.ZeroInt()), err
+	}
+
+	// At most we can swap half of the liquidity in the pool
+	liquidTokenAmt := liquidity.AmountOf(outputCoin.Denom).Quo(sdk.NewInt(2))
```

Comment:
Is this just a self imposed limit?

Reply:
Yea, this would allow us to still capture some arb on illiquid pools while ensuring we never error out when we make the calculation on upper bound. Additionally, on larger/more liquid pools, the trade size will never be larger than `ExtendedMaxInputAmount` which most certainly will be less than half the liquidity in a liquid pool.